casanovo/data/db_utils.py

```diff
@@ -0,0 +1,389 @@
+        candidates.sort_values(by=["calc_mass", "peptide"], inplace=True)
+        return candidates["peptide"], candidates["protein"]
+
+    def get_associated_protein(self, peptide: str) -> str:
+        """
+        Returns the associated protein for a given peptide.
+
+        Parameters
+        ----------
+        peptide : str
+            The peptide sequence.
+
+        Returns
+        -------
+        protein : str
+            The associated protein(s).
+        """
+        return ",".join(self.prot_map[peptide])
+
+    def _digest_fasta(
+        self,
+        fasta_filename: str,
+        enzyme: str,
+        digestion: str,
+        missed_cleavages: int,
+        max_mods: int,
+        min_peptide_length: int,
+        max_peptide_length: int,
+    ) -> pd.DataFrame:
+        """
+        Digests a FASTA file and returns the peptides, their masses, and associated protein.
+
+        Parameters
+        ----------
+        fasta_filename : str
+            Path to the FASTA file.
+        enzyme : str
+            The enzyme to use for digestion.
+            See pyteomics.parser.expasy_rules for valid enzymes.
+            Can also be a regex pattern.
+        digestion : str
+            The type of digestion to perform. Either 'full', 'partial' or 'non-specific'.
+        missed_cleavages : int
+            The number of missed cleavages to allow.
+        max_mods : int
+            The maximum number of modifications to allow per peptide.
+        min_peptide_length : int
+            The minimum length of peptides to consider.
+        max_peptide_length : int
+            The maximum length of peptides to consider.
+
+        Returns
+        -------
+        pep_table : pd.DataFrame
+            A Pandas DataFrame with peptide, mass,
+            and protein columns. Sorted by neutral mass in ascending order.
+        prot_map : dict
+            A dictionary mapping peptides to associated proteins.
+        """
+        # Verify the existence of the file:
+        if not os.path.isfile(fasta_filename):
+            logger.error("File %s does not exist.", fasta_filename)
+            raise FileNotFoundError(f"File {fasta_filename} does not exist.")
+
+        peptide_list = []
+        if digestion not in ["full", "partial", "non-specific"]:
+            logger.error("Digestion type %s not recognized.", digestion)
+            raise ValueError(f"Digestion type {digestion} not recognized.")
+        if enzyme not in parser.expasy_rules:
+            logger.info(
+                "Enzyme %s not recognized. Interpreting as cleavage rule.",
+                enzyme,
+            )
+        valid_aa = set(list(self.residues.keys()) + ["C"])
+        if digestion == "non-specific":
+            for header, seq in fasta.read(fasta_filename):
+                pep_set = []
+                # Generate all possible peptides
+                for i in range(len(seq)):
+                    for j in range(i + 1, len(seq) + 1):
```

Comment:
**todo:** This can be optimized by making the end of the range take `min_peptide_len` and `max_peptide_len` already into account:
`range(i + min_peptide_len, min(i + max_peptide_len + 1, len(seq) + 1)`.